Content.Shared/_DV/CosmicCult/SharedCosmicCultSystem.cs

```diff
@@ -0,0 +1,93 @@
+using Content.Shared._DV.CosmicCult.Components;
+using Content.Shared.Antag;
+using Content.Shared.Ghost;
+using Content.Shared.Mind;
+using Content.Shared.Roles;
+using Content.Shared._DV.Roles;
+using Robust.Shared.GameStates;
+using Robust.Shared.Player;
+
+namespace Content.Shared._DV.CosmicCult;
+
+public abstract class SharedCosmicCultSystem : EntitySystem
+{
+    [Dependency] private readonly SharedMindSystem _mind = default!;
+    [Dependency] private readonly SharedRoleSystem _role = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<CosmicCultComponent, ComponentGetStateAttemptEvent>(OnCosmicCultCompGetStateAttempt);
+        SubscribeLocalEvent<CosmicCultLeadComponent, ComponentGetStateAttemptEvent>(OnCosmicCultCompGetStateAttempt);
+        SubscribeLocalEvent<CosmicCultComponent, ComponentStartup>(DirtyCosmicCultComps);
+        SubscribeLocalEvent<CosmicCultLeadComponent, ComponentStartup>(DirtyCosmicCultComps);
+    }
+
+    public bool EntityIsCultist(EntityUid user)
+    {
+        if (!_mind.TryGetMind(user, out var mind, out _))
+            return false;
+
+        return HasComp<CosmicCultComponent>(user) || _role.MindHasRole<CosmicCultRoleComponent>(mind);
+    }
+
+    public bool EntitySeesCult(EntityUid user)
+    {
+        return EntityIsCultist(user) || HasComp<GhostComponent>(user);
+    }
+
+    /// <summary>
+    /// Determines if a Cosmic Cult Lead component should be sent to the client.
+    /// </summary>
+    private void OnCosmicCultCompGetStateAttempt(EntityUid uid, CosmicCultLeadComponent comp, ref ComponentGetStateAttemptEvent args)
+    {
+        args.Cancelled = !CanGetState(args.Player);
+    }
+
+    /// <summary>
+    /// Determines if a Cosmic Cultist component should be sent to the client.
+    /// </summary>
+    private void OnCosmicCultCompGetStateAttempt(EntityUid uid, CosmicCultComponent comp, ref ComponentGetStateAttemptEvent args)
+    {
+        args.Cancelled = !CanGetState(args.Player);
+    }
+
+    /// <summary>
+    /// The criteria that determine whether a Cult Member component should be sent to a client.
+    /// </summary>
+    /// <param name="player">The Player the component will be sent to.</param>
+    private bool CanGetState(ICommonSession? player)
+    {
+        //Apparently this can be null in replays so I am just returning true.
+        if (player?.AttachedEntity is not { } uid)
+            return true;
+
+        if (EntitySeesCult(uid) || HasComp<CosmicCultComponent>(uid) || HasComp<CosmicCultLeadComponent>(uid))
```

Comment:
Redundant with `EntitySeesCult`
```suggestion
        if (EntitySeesCult(uid) || HasComp<CosmicCultLeadComponent>(uid))
```